Fix related to UpdateProgramApi

Azher2Ali opened the pull request:
<img width="1076" alt="Screenshot 2024-03-01 at 11 37 52 AM" src="https://github.com/icgc-argo/program-service/assets/121898125/254c984c-7f47-4edf-a98a-e3e4dc5b2b28">
<img width="1061" alt="Screenshot 2024-03-01 at 11 38 00 AM" src="https://github.com/icgc-argo/program-service/assets/121898125/302c7a94-cc8f-4e50-a251-16f795b02d4d">
<img width="1069" alt="Screenshot 2024-03-01 at 11 38 14 AM" src="https://github.com/icgc-argo/program-service/assets/121898125/a6378f8f-13b6-477c-89d3-18dbad2ebb40">
As there was issue in the return response of UpdateProgramApi, we found out that the mapping between proto generated files and REST pojo's are failing on MembershipType enum. In order to fix this we have written custom mapping and used that for transformation between the two.

As shown in screenshots, we have tested the UpdateProgramApi in local by changes its fields which returns the response as expected and those changes are reflected in getProgramApi.

demariadaniel commented on src/main/java/org/icgc/argo/program_service/converter/Grpc2JsonConverter.java:
```diff
@@ -49,7 +52,30 @@ public String getJsonFromObject(Object response) {
+  public UpdateProgramRequest prepareUpdateProgramRequest (UpdateProgramRequestDTO updateProgramRequestDTO) {
+
+    Program program = Program.newBuilder()
+            .addAllCancerTypes(updateProgramRequestDTO.getProgram().getCancerTypes())
```
Not a Java dev so forgive my ignorance, but I'm seeing a lot of `updateProgramRequestDTO.getProgram()`, could this be more DRY, like captured in a variable once?

Azher2Ali replied:
Yes, you're right! We can capture the result of updateProgramRequestDTO.getProgram() in a variable to make the code more DRY. Will update it accordingly

joneubank commented on src/main/java/org/icgc/argo/program_service/services/ProgramServiceFacade.java:
```diff
@@ -173,13 +173,34 @@ public UpdateProgramResponse updateProgram(UpdateProgramRequest request) {
   }
 
+  @Transactional
+  public UpdateProgramResponse updateProgramWithDataCenter (UpdateProgramRequest request, DataCenterDetailsDTO dataCenterDetailsDTO) {
```
Can this new function just take the UpdateProgramDTO instead of relying on the GRPC message type? We want to eventually remove all uses of the GRPC code.

joneubank commented on src/main/resources/application.yml:
```diff
@@ -12,7 +12,7 @@ spring:
     username: postgres
     password: password
   mail:
-    host: mailhog.qa.argo.cancercollaboratory.org
```
Ideally we would replace this with a generic URL instead of hardcoding a specific test environment.

joneubank commented on src/main/resources/application.yml:
```diff
@@ -32,7 +32,7 @@ retry:
     multiplier: 2.0
 
 app:
-  egoUrl: "https://ego.qa.argo.cancercollaboratory.org/api"
```
Same comment about not hardcoded default environments.

joneubank commented on src/main/java/org/icgc/argo/program_service/converter/Grpc2JsonConverter.java:
```diff
@@ -49,7 +52,30 @@ public String getJsonFromObject(Object response) {
     }
   }
-
+  public UpdateProgramRequest prepareUpdateProgramRequest (UpdateProgramRequestDTO updateProgramRequestDTO) {
```
So this looks like we're mixing up our GRPC messages (UpdateProgramRequest) with our DTOs from the HTTP endpoints. I mention in the comment on the `updateProgramWithDataCenter` function that it is best not to involve the GRPC objects in our service logic... If so, then hopefully this whole function can be skipped.

joneubank reviewed (changes requested):
Just want to check if we can avoid using the GRPC messages and types in our services and our HTTP controllers. We want to be removing the GRPC code (including all protos) as soon as the HTTP controllers are working end to end.

Azher2Ali commented:
As per the review, we have made changes which are related to removing the gRPC dependent code from the update API implementation along with the defect fix where we have modified the mapping so that all the fields are mapped properly during update

UmmulkiramR commented on src/main/java/org/icgc/argo/program_service/services/ProgramService.java:
dataCenterDetailsDTO.getId().isEmpty() does not check for null strings. Can we be sure getId() is never going to return a null?

leoraba commented on src/main/java/org/icgc/argo/program_service/services/ProgramService.java:
```diff
@@ -283,6 +286,7 @@ private List<InstitutionEntity> filterAndAddInstitutions(@NonNull List<String> n
   public ProgramEntity updateProgram(
       @NonNull ProgramEntity programToUpdate,
       @NonNull ProgramEntity updatingProgram,
+      @NotNull DataCenterDetailsDTO dataCenterDetailsDTO,
```
Confusing why annotating dataCenterDetailsDTO as NotNull, and in the facade we're still passing this value as null.

leoraba replied:
Also notice this param is using `NotNull` and other params are using 'NonNull' from Lombok.

leoraba replied:
I'd suggest removing this annotation if you really want to pass a null dataCenter value, as I see you have already an If condition to check null dataCenterDetailsDTO

leoraba commented on src/main/java/org/icgc/argo/program_service/services/ProgramService.java:
```diff
@@ -491,6 +504,15 @@ public DataCenterEntity updateDataCenter(
   }
 
+  public void processDataCenter (
+          @NonNull DataCenterDetailsDTO dataCenterDetailsDTO, @NotNull DataCenterEntity updatingDataCenter) {
```
same here. Noticing a `NonNull` vs `NotNull`

Azher2Ali commented:
**As per the feedback, the recent PR holds changes to:**
1. Null handling for datacenterId- datacenterId can now be null under certain conditions
2. Removed NotNull annotation- The @NotNull annotation, which typically ensures that a field cannot be null, has been removed from the codebase.
3. Created overridden method for update program- A new method has been added to the codebase, likely within a class related to updating programs
4. Decoupled gRPC code from Rest code:  This refers to a separation/restructuring of the codebase to reduce dependencies between gRPC and REST components.

Azher2Ali commented:
With the transformation from CancerCollaboratory to Cumulus, there has been changes to default environments and in order fix the test case failures related to ProgramServiceImplTest we have made the recent commit in which we have updated the AppProperties.java by annotating the getEgoPublicKey method to not be initiated while running tests. Also, we have removed the SpringDataSource properties from the test profile in application.yml so that it picks the values from the default profile as all other profiles.

joneubank commented on src/main/java/org/icgc/argo/program_service/properties/AppProperties.java:
```diff
@@ -121,6 +121,7 @@ public RestTemplate RestTemplate() {
   }
 
   @Bean
+  @Profile("!test")
```
👍

joneubank commented on src/main/resources/application.yml:
```diff
@@ -88,11 +88,11 @@ spring.mail:
-  url: jdbc:tc:postgresql:9.5.13://localhost:5432/program_db?TC_INITFUNCTION=org.icgc.argo.program_service.test.FlywayInit::initTestContainers
-  username: postgres
-  password:
+#spring.datasource:
+#  driver-class-name: org.testcontainers.jdbc.ContainerDatabaseDriver
+#  url: jdbc:tc:postgresql:9.5.13://localhost:5432/program_db?TC_INITFUNCTION=org.icgc.argo.program_service.test.FlywayInit::initTestContainers
+#  username: postgres
+#  password:
```
Can we just remove the unused config instead of commenting?

joneubank reviewed (changes requested):
This is a good fix. Let's just remove those comments and then can be merged. Good to see tests passing again!
